### PR TITLE
Robust XML assertions

### PR DESCRIPTION
**Pull request by eNeRGy164**
More coverage of #1039

* Added failure message when Subject is null
* Added Guards against null parameters

## IMPORTANT 

* [x] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [x] The changes are covered by a new or existing set of unit tests which follow the Arrange-Act-Assert syntax such as is used [in this example](https://github.com/fluentassertions/fluentassertions/blob/daaf35b9b59b622c96d0c034e8972a020b2bee55/Tests/FluentAssertions.Shared.Specs/BasicEquivalencySpecs.cs#L33).
* [x] If the contribution adds a feature or fixes a bug, please update [**the release notes**](https://github.com/fluentassertions/fluentassertions/tree/master/docs), which are published on the [website](https://fluentassertions.com/releases).

### CODE REVIEW

File: Src/FluentAssertions/Xml/XAttributeAssertions.cs [hunk 1]
```diff
@@ -30,12 +32,23 @@ public XAttributeAssertions(XAttribute attribute)
         /// <param name="becauseArgs">
         /// Zero or more objects to format using the placeholders in <paramref name="because" />.
         /// </param>
+        /// <exception cref="ArgumentNullException"><paramref name="expected"/> is <c>null</c>.</exception>
         public AndConstraint<XAttributeAssertions> Be(XAttribute expected, string because = "", params object[] becauseArgs)
         {
-            Execute.Assertion
-                .ForCondition(Subject.Name == expected.Name && Subject.Value == expected.Value)
+            Guard.ThrowIfArgumentIsNull(expected, nameof(expected));
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Xml/XAttributeAssertions.cs:38]: 🤔 I wonder about this one. What if you're comparing two references to an `XAttribute` and both are `null`. Shouldn't this assertion succeed? This is also how `StringAssertions` work.

Comment (eNeRGy164) [reply Src/FluentAssertions/Xml/XAttributeAssertions.cs:38]: Yeah, that might be a good exception.

Comment (jnyrup) [reply Src/FluentAssertions/Xml/XAttributeAssertions.cs:38]: And the same applies for `[Not]BeEquivalentTo`

File: Tests/FluentAssertions.Specs/Xml/XAttributeAssertionSpecs.cs [hunk 2]
```diff
@@ -40,6 +40,35 @@ public void When_asserting_an_xml_attribute_is_equal_to_a_different_xml_attribut
...
+                theAttribute.Should().Be(new XAttribute("name", "value"), "we want to test the failure {0}", "message");
+
+            // Assert
+            act.Should().Throw<XunitException>()
+                .WithMessage(
+                    "Expected the attribute to be name=\"value\" *failure message*, but theAttribute is <null>.");
+        }
+
+        [Fact]
+        public void When_asserting_àttribute_should_be_equal_to_null_it_should_throw()
```
Comment (lg2de) [inline Tests/FluentAssertions.Specs/Xml/XAttributeAssertionSpecs.cs:59]: You added (accidentally?) an `à` instead of `a`.

File: Tests/FluentAssertions.Specs/Xml/XAttributeAssertionSpecs.cs [hunk 3]
```diff
@@ -87,6 +116,35 @@ public void When_asserting_an_xml_attribute_is_not_equal_to_the_same_xml_attribu
...
+                theAttribute.Should().NotBe(new XAttribute("name", "value"), "we want to test the failure {0}", "message");
+
+            // Assert
+            act.Should().Throw<XunitException>()
+                .WithMessage(
+                    "Did not expect the attribute to be name=\"value\" *failure message*, but theAttribute is <null>.");
+        }
+
+        [Fact]
+        public void When_asserting_àttribute_should_be_different_to_null_it_should_throw()
```
Comment (lg2de) [inline Tests/FluentAssertions.Specs/Xml/XAttributeAssertionSpecs.cs:135]: Another `à`.

File: Tests/FluentAssertions.Specs/Xml/XAttributeAssertionSpecs.cs (hunk 2 above)
Comment (eNeRGy164) [reply Tests/FluentAssertions.Specs/Xml/XAttributeAssertionSpecs.cs:59]: Yeah, must have hit it while coding in bed :)

File: Src/FluentAssertions/Xml/XAttributeAssertions.cs (hunk 1 above)
Comment (eNeRGy164) [reply Src/FluentAssertions/Xml/XAttributeAssertions.cs]: It was more work then expected.
Also added more tests to cover these scenarios.

File: Src/FluentAssertions/Xml/Equivalency/XmlReaderValidator.cs [hunk 4]
```diff
@@ -209,5 +221,17 @@ private Failure ValidateText()
 
             return null;
         }
+
+        private Failure ValidateAgainstNulls()
+        {
+            if ((expectationIterator is null) ^ (subjectIterator is null))
```
Comment (jnyrup) [inline Src/FluentAssertions/Xml/Equivalency/XmlReaderValidator.cs:227]: When trying this out with
```cs
var document = null as XDocument;
var otherDocument = XDocument.Parse("<parent><child /></parent>");
document.Should().BeEquivalentTo(otherDocument);
```

I get a somewhat unhelpful failure message.
```
Expected document to be equivalent to <null>, but found "".
```

Instead of adding checks in `XmlReaderValidator`, I would prefer if we null checked at call-site.
It's more lines of code, but it avoids adding complexity to `XmlReaderValidator`.

Something like this
```cs
public AndConstraint<XDocumentAssertions> BeEquivalentTo(XDocument expected, string because = "", params object[] becauseArgs)
{
    if (Subject is null && expected is null)
    {
        return new AndConstraint<XDocumentAssertions>(this);
    }

    bool success = Execute.Assertion
        .BecauseOf(because, becauseArgs)
        .ForCondition(Subject is not null && expected is not null)
        .FailWith(() => new FailReason("Expected {context:subject} to be equivalent to {0}{reason}, but found {1}.",
            Subject?.ToString(), expected?.ToString()));

    if (success)
    {
        using (XmlReader subjectReader = Subject.CreateReader())
        using (XmlReader otherReader = expected.CreateReader())
        {
            var xmlReaderValidator = new XmlReaderValidator(subjectReader, otherReader, because, becauseArgs);
            xmlReaderValidator.Validate(shouldBeEquivalent: true);
        }
    }

    return new AndConstraint<XDocumentAssertions>(this);
}
```

The code above gives this failure message.
```
Expected document to be equivalent to <null>, but found "<parent>
      <child />
    </parent>".
```

Comment (eNeRGy164) [reply Src/FluentAssertions/Xml/Equivalency/XmlReaderValidator.cs:227]: Why would `XmlReaderValidator` not handle null values on its own? I took this code from `StringEqualityValidator` (`StringValidator`) where it is responsible for the null checks.

Comment (jnyrup) [reply Src/FluentAssertions/Xml/Equivalency/XmlReaderValidator.cs:227]: As `StringValidator` also does it that way I'm fine with it and your solution solves the problem 👍 
It's internal so we can always improve later if we want to.

Comment (eNeRGy164) [reply Src/FluentAssertions/Xml/Equivalency/XmlReaderValidator.cs]: The correct content is now also part of the error message.

**Review by jnyrup — APPROVED** [review]

File: Src/FluentAssertions/Xml/Equivalency/XmlReaderValidator.cs [hunk 5]
```diff
@@ -209,5 +226,27 @@ private Failure ValidateText()
 
             return null;
         }
+
+        private Failure ValidateAgainstNulls()
+        {
+            if ((expectationReader is null) ^ (subjectReader is null))
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Xml/Equivalency/XmlReaderValidator.cs:232]: 🤔 I think it would be better to use more ordinary `&&`, `||` and negation instead of using xor. I had to look up this operator to remind myself what it did again.

File: Src/FluentAssertions/Xml/Equivalency/XmlReaderValidator.cs [hunk 6]
```diff
@@ -209,5 +226,27 @@ private Failure ValidateText()
...
+            {
+                return new Failure(
+                    "Expected {context:subject} to be equivalent to {0}{reason}, but found {1}.",
+                    GetReaderContent(subjectReader), GetReaderContent(expectationReader));
+            }
+
+            return null;
+
+            static string GetReaderContent(XmlReader reader)
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Xml/Equivalency/XmlReaderValidator.cs:241]: 🤔 Isn't it cleaner to create a little `IValueFormatter` for XML readers?

File: Src/FluentAssertions/Xml/Equivalency/XmlReaderValidator.cs [hunk 7]
```diff
@@ -38,6 +40,21 @@ public void Validate(bool shouldBeEquivalent)
 
         private Failure Validate()
         {
+            if (subjectReader is null && expectationReader is null)
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Xml/Equivalency/XmlReaderValidator.cs:43]: 🤔 It's kind of weird that this block and `ValidateAgainstNull` are both testing against `null`. Shouldn't this be combined?

File: Src/FluentAssertions/Xml/Equivalency/XmlReaderValidator.cs (hunk 5 above)
Comment (eNeRGy164) [reply Src/FluentAssertions/Xml/Equivalency/XmlReaderValidator.cs:232]: 🤡 I Just copied it from https://github.com/fluentassertions/fluentassertions/blob/develop/Src/FluentAssertions/Primitives/StringValidator.cs

File: Src/FluentAssertions/Xml/Equivalency/XmlReaderValidator.cs (hunk 7 above)
Comment (eNeRGy164) [reply Src/FluentAssertions/Xml/Equivalency/XmlReaderValidator.cs:43]: 🤡 I Just copied it from https://github.com/fluentassertions/fluentassertions/blob/develop/Src/FluentAssertions/Primitives/StringValidator.cs

File: Src/FluentAssertions/Xml/Equivalency/XmlReaderValidator.cs (hunk 6 above)
Comment (eNeRGy164) [reply Src/FluentAssertions/Xml/Equivalency/XmlReaderValidator.cs:241]: Ah, didn't think of that

File: Tests/FluentAssertions.Specs/Xml/XDocumentAssertionSpecs.cs [hunk 8]
```diff
@@ -123,35 +131,43 @@ public void When_asserting_a_xml_document_is_not_equal_to_the_same_xml_document_
         }
 
         [Fact]
-        [SuppressMessage("ReSharper", "ExpressionIsAlwaysNull")]
```
[the remark this answers is not in the document]
Comment (eNeRGy164) [reply Tests/FluentAssertions.Specs/Xml/XDocumentAssertionSpecs.cs:126]: I'm not sure what to do here as I have no Rider to confirm if the behavior is stall present.
Visual Studio is actively saying the attribute is redundant, and I also would expect this message then on every unit test with `null`'s and this is the only place in the past weeks where I encountered these attributes.

[the remark this answers is not in the document]
Comment (jnyrup) [reply Tests/FluentAssertions.Specs/Xml/XDocumentAssertionSpecs.cs:126]: I have some other analyzers to adjust, so I'll take of this as well.
So the PR is fine.

File: Src/FluentAssertions/Xml/Equivalency/XmlReaderValidator.cs (hunk 5 above)
Comment (jnyrup) [reply Src/FluentAssertions/Xml/Equivalency/XmlReaderValidator.cs:232]: It's only 20 months since you looked it up last time 🙃 
https://github.com/fluentassertions/fluentassertions/pull/1141#discussion_r326902258

Would changing it to 
```cs
if ((expectationReader is null) != (subjectReader is null))
```
be more readable to you?
It's an equivalent way to expresses "one, but not both".

Comment (dennisdoomen) [reply Src/FluentAssertions/Xml/Equivalency/XmlReaderValidator.cs:232]: 20 months?? I don't even remember what I did last week ;-). 

But yes, much more expressive.

Comment (jnyrup) [reply Src/FluentAssertions/Xml/Equivalency/XmlReaderValidator.cs:232]: FYI, I've changed the other occurences of `^` to `!=` in #1574.

Comment (eNeRGy164) [reply Src/FluentAssertions/Xml/Equivalency/XmlReaderValidator.cs]: Changed

File: Src/FluentAssertions/Xml/Equivalency/XmlReaderValidator.cs (hunk 7 above)
Comment (eNeRGy164) [reply Src/FluentAssertions/Xml/Equivalency/XmlReaderValidator.cs:43]: You can't combine it, as it is a tri-state:

1. Both `null` -> Stop validation without failure
2. One `null` -> Stop validation with failure
3. None `null` -> Continue validation

Comment (dennisdoomen) [reply Src/FluentAssertions/Xml/Equivalency/XmlReaderValidator.cs:43]: Fair enough

**Review by dennisdoomen — APPROVED** [review]
